tests/test_simpleButler.py

```diff
@@ -882,10 +883,75 @@ def makeButler(self, writeable: bool = False) -> Butler:
+        ref = butler.find_dataset("flat", detector=2, physical_filter="Cam1-G")
+        self.assertIsInstance(ref, DatasetRef)
+
+        # Get the butler root for the URI.
+        config_dir = butler._config["root"]
+
+        # Read it via a repo label and a path.
+        with tempfile.NamedTemporaryFile(mode="w", suffix=".yaml") as index_file:
+            label = "test_repo"
+            index_file.write(f"{label}: {config_dir}\n")
+            index_file.flush()
+            with mock_env({"DAF_BUTLER_REPOSITORY_INDEX": index_file.name}):
+                butler_factory = LabeledButlerFactory()
+                factory = butler_factory.bind(access_token=None)
+
+                for dataset_uri in (
+                    f"ivo://org.rubinobs/usdac/test?repo={config_dir}&id={ref.id}",
+                    f"ivo://org.rubinobs/ukdac/lsst-dr1?repo={config_dir}/butler.yaml&id={ref.id}",
+                    f"butler://{label}/{ref.id}",
+                    f"ivo://org.rubinobs/usdac/lsst-dp1?repo={label}&id={ref.id}",
+                ):
+                    new_butler, ref2 = Butler.get_dataset_from_uri(dataset_uri)
+                    self.assertEqual(ref, ref2)
+                    # The returned butler needs to have the datastore mocked.
+                    DatastoreMock.apply(new_butler)
+                    dataset_id, _ = butler.get(ref2)
+                    self.assertEqual(dataset_id, ref.id)
+
+                    factory_butler, ref2 = Butler.get_dataset_from_uri(dataset_uri, factory=factory)
+                    self.assertEqual(ref, ref2)
+                    # The returned butler needs to have the datastore mocked.
+                    DatastoreMock.apply(factory_butler)
+                    dataset_id, _ = factory_butler.get(ref2)
+                    self.assertEqual(dataset_id, ref.id)
+
+                # Non existent dataset.
+                missing_id = str(ref.id).replace("2", "3")
+                _, no_ref = Butler.get_dataset_from_uri(f"butler://{label}/{missing_id}")
+                self.assertIsNone(no_ref)
+
+        # Test some failure modes.
+        for dataset_uri in (
+            "butler://label/1234",  # Bad UUID.
+            "butler://1234",  # No UUID.
+            "butler:///1234",  # No label.
+            "ivo://rubin/1234",  # No query part and bad UUID and no label.
+            "ivo://rubin/datasets/dr1/82d79caa-0823-4300-9874-67b737367ee0",  # No query part.
+            "ivo://org.rubinobs/datasets?repo=dr1&id=1234",  # Bad UUID.
+            "ivo://org.rubinobs/butler?release=dr1&id=82d79caa-0823-4300-9874-67b737367ee0",  # No repo key.
+            "ivo://org.rubinobs/butler?repo=dr1&repo=dr2&id=82d79caa-0823-4300-9874-67b737367ee0",  # 2 vals.
+            "ivo://org.rubinobs/something?repo=%20&id=82d79caa-0823-4300-9874-67b737367ee0",  # no repo.
+            "https://something.edu/1234",  # Wrong scheme.
+        ):
```

Comment:
Might it also be useful to test percent-encoded params?

Reply:
The `%20` is converted to a space for the failing test here but I added a `%2F` for a `/` in the URIs that work.